File: digidis-go/day19/day19.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
	"time"
)

var (
	reg     = []int{0, 0, 0, 0, 0, 0, 0}
	code    [][]int
	ipr     int
	opcodes = make(map[int]string)
)

func main() {
	start := time.Now()
	defer func() {
		fmt.Printf("Done in %v\n", time.Since(start))
	}()

	data, _ := ioutil.ReadFile("input.txt")
	rows := strings.Split(string(data), "\n")

	for _, r := range rows {
		c := strings.Fields(r)
		if c[0] == "#ip" {
			ipr = atoi(c[1])
			continue
		}
		code = append(code, []int{
			instructions[c[0]], atoi(c[1]), atoi(c[2]), atoi(c[3]),
		})
	}

	v, o := run(0)
	fmt.Printf("Part 1: %v (%d operations)\n", v, o)

	v, o = run(1)
	fmt.Printf("Part 2: %v (%d operations)\n", v, o)
}

func run(r0 int) (int, int) {
	reg = []int{r0, 0, 0, 0, 0, 0, 0}
	ip := 0
	ops := 0
	for {
		reg[ipr] = ip
		eval(code[ip])
		ip = reg[ipr] + 1
		if ip >= len(code) {
			break
		}
		ops++
	}
	return reg[0], ops
}

var instructions = map[string]int{
	"addr": 0, "addi": 1, "mulr": 2, "muli": 3,
	"banr": 4, "bani": 5, "borr": 6, "bori": 7,
	"setr": 8, "seti": 9, "gtir": 10, "gtri": 11,
	"gtrr": 12, "eqir": 13, "eqri": 14, "eqrr": 15,
}

func eval(code []int) {
	switch code[0] {
	case 0:
		reg[code[3]] = reg[code[1]] + reg[code[2]]
	case 1:
		reg[code[3]] = reg[code[1]] + code[2]
	case 2:
		reg[code[3]] = reg[code[1]] * reg[code[2]]
	case 3:
		reg[code[3]] = reg[code[1]] * code[2]
	case 4:
		reg[code[3]] = reg[code[1]] & reg[code[2]]
	case 5:
		reg[code[3]] = reg[code[1]] & code[2]
	case 6:
		reg[code[3]] = reg[code[1]] | reg[code[2]]
	case 7:
		reg[code[3]] = reg[code[1]] | code[2]
	case 8:
		reg[code[3]] = reg[code[1]]
	case 9:
		reg[code[3]] = code[1]
	case 10:
		if code[1] > reg[code[2]] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	case 11:
		if reg[code[1]] > code[2] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	case 12:
		if reg[code[1]] > reg[code[2]] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	case 13:
		if code[1] == reg[code[2]] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	case 14:
		if reg[code[1]] == code[2] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	case 15:
		if reg[code[1]] == reg[code[2]] {
			reg[code[3]] = 1
		} else {
			reg[code[3]] = 0
		}
	}
}

func atoi(s string) int {
	i, _ := strconv.Atoi(s)
	return i
}

File: erikdyrelius-python/day19.py
from aocbase import readInput
import re
from collections import deque

inp = readInput()
p = re.compile(r"(\w+) (\d+) (\d+) (\d+)")
inp2 = '''#ip 0
seti 5 0 1
seti 6 0 2
addi 0 1 0
addr 1 2 3
setr 1 0 0
seti 8 0 4
seti 9 0 5'''


def doAddr(inst, reg):
    reg[inst[3]] = reg[inst[1]] + reg[inst[2]]

def doAddi(inst, reg):
    reg[inst[3]] = reg[inst[1]] + inst[2]

def doMulr(inst, reg):
    reg[inst[3]] = reg[inst[1]] * reg[inst[2]]

def doMuli(inst, reg):
    reg[inst[3]] = reg[inst[1]] * inst[2]

def doBanr(inst, reg):
    reg[inst[3]] = reg[inst[1]] & reg[inst[2]]

def doBani(inst, reg):
    reg[inst[3]] = reg[inst[1]] & inst[2]

def doBorr(inst, reg):
    reg[inst[3]] = reg[inst[1]] | reg[inst[2]]

def doBori(inst, reg):
    reg[inst[3]] = reg[inst[1]] | inst[2]

def doSetr(inst, reg):
    reg[inst[3]] = reg[inst[1]]

def doSeti(inst, reg):
    reg[inst[3]] = inst[1]

def doGtir(inst, reg):
    if inst[1] > reg[inst[2]]:
        reg[inst[3]] = 1
    else:
        reg[inst[3]] = 0

def doGtri(inst, reg):
    if reg[inst[1]] > inst[2]:
        reg[inst[3]] = 1
    else:
        reg[inst[3]] = 0

def doGtrr(inst, reg):
    if reg[inst[1]] > reg[inst[2]]:
        reg[inst[3]] = 1
    else:
        reg[inst[3]] = 0

def doEqir(inst, reg):
    if inst[1] == reg[inst[2]]:
        reg[inst[3]] = 1
    else:
        reg[inst[3]] = 0

def doEqri(inst, reg):
    if reg[inst[1]] == inst[2]:
        reg[inst[3]] = 1
    else:
        reg[inst[3]] = 0

def doEqrr(inst, reg):
    if reg[inst[1]] == reg[inst[2]]:
        reg[inst[3]] = 1
    else:
        reg[inst[3]] = 0

d = {
    "addr":doAddr,
    "addi":doAddi,
    "mulr":doMulr,
    "muli":doMuli,
    "banr":doBanr,
    "bani":doBani,
    "borr":doBorr,
    "bori":doBori,
    "setr":doSetr,
    "seti":doSeti,
    "gtir":doGtir,
    "gtri":doGtri,
    "gtrr":doGtrr,
    "eqir":doEqir,
    "eqri":doEqri,
    "eqrr":doEqrr
}

def parse(s):
    l = []
    for line in s.splitlines():
        m = p.match(line)
        if m:
            l.append([d[m.group(1)]] + list(map(int, m.groups()[1:])))
        else:
            if line[0]=='#':
                ip = int(line[4])
            else:
                print("Error:"+line)
    return l, ip

def run(ip, prog, reg=[0]*6):
    while True:
        instr = prog[reg[ip]]
        print(reg, instr)
        instr[0](instr, reg)
        reg[ip] += 1
        if reg[ip] < 0 or reg[ip] >= len(prog):
            break
    return reg[0]


instr, ip = parse(inp)
print(run(ip, instr))
print(run(ip, instr, [1, 0, 0, 0, 0, 0]))

print("Solution to day 19 part 1:",19.1)
print("Solution to day 19 part 2:",19.2)
